### DISCUSSION TITLE
element property by the SnapshotDb

### DISCUSSION THREAD

**Discussion by paprophet** [iTwin/itwinjs-core#493] [Q&A] [answered]
i use SimplePropertyGrid and can display the element property ,but i want  select all elements property by the SnapshotDb in the end console.

**Comment by pmconne** [comment]
Can you please clarify your question? I don't understand what you're trying to achieve.

**Reply by paprophet** [reply]
in  Command Line Applications (https://medium.com/itwinjs/writing-command-line-applications-with-imodel-js-ba06f4531cc2) ,
    const _snapshotdb = SnapshotDb.openFile(bimfile);
   let elementPropertys=   _snapshotdb.elements.getElementProps('0x324');

i get the element property have two problem:
1.the element propertys donot have the categories/group 
2.the element some property is id but not the value. 

but in the frontend code ,in the PropertyGrid.tsx ( 
propertyData = await this.props.dataProvider.getData();
)
the propertyData  is i want result,
so how i can get the same result by Command Line Applications Code!

**Comment by pmconne** [comment]
`IModelDb.Elements.getElementProps` gives you only the properties of a specific element - i.e., the properties that are stored directly on that element.
`PropertyGrid.gatherData` uses the EC presentation system to collect properties from potentially multiple elements and organize them into categories.
@grigasp may be able to tell you how to do the latter on the backend.

**Comment by grigasp** [comment, marked as answer]
The data provider interface on the frontend is specific to our React-based UI layer. The `PresentationPropertyDataProvider` implementation uses the `Presentation` API to retrieve the properties. As @pmconne already summarized, the presentation system collects properties of the element and its related elements and aspects.

The data provider interface is not available on the backend - you have to use the Presentation API directly. An example of that could be this (a little outdated) gist: https://gist.github.com/grigasp/aa75511dc4137137b651f56676c27f87. Have a look at `listElementProperties` function there. 

Additionally, you might find this sample useful: https://www.itwinjs.org/sample-showcase/?group=Viewer+Features&sample=property-formatting-sample&imodel=Retail+Building+Sample, especially the "3. Do it yourself" approach.

**Reply by paprophet** [reply]
in fact, i submit the bim file to my local server, then  write all  element property(the data like frontend ) to my database (mongo db ),and 
create diffent element tree by the diffent property value from my database,so,
i want read all element property by SnapshotDb.openFile(bimfile) and run ecsql or other method  on the backend
so,i want imodeljs can give me the methods not open the frontend. or imodeljs next version can do this, 
thank you !

**Reply by grigasp** [reply]
There are Presentation APIs on both sides - backend and frontend. The [gist I linked](https://gist.github.com/grigasp/aa75511dc4137137b651f56676c27f87) is actually a backend-only example. However, you can use frontend examples as well, since APIs on both sides are very similar.

**Reply by paprophet** [reply]
thank you !
but there have 2 problem:
1.the Presentation.getManager().computeSelection(...) method is  deprecated in new imodeljs version(2.10.0),so i want what method can replace of this old method? 

2.when i set one elementid in this method ,the return value is right,but when i set elementarray in this method,the return value is not right(the return value is  the same property in the element array, lost the diffent property)
so now i must set the element id one by one and read the property is slow .

**Reply by grigasp** [reply]
1. There are two overrides of that function - [one is deprecated](https://github.com/imodeljs/imodeljs/blob/2ecb722f217ffdbcaa258fef2282fe6b4b58a89d/presentation/backend/src/presentation-backend/PresentationManager.ts#L811) and [the other is not](https://github.com/imodeljs/imodeljs/blob/2ecb722f217ffdbcaa258fef2282fe6b4b58a89d/presentation/backend/src/presentation-backend/PresentationManager.ts#L816).

2. The example passes the `ContentFlags.MergeResults` flag [here](https://gist.github.com/grigasp/aa75511dc4137137b651f56676c27f87#file-main-ts-L90), which, when selecting properties of multiple elements, merges values into one record. With that flag you always get one item in the resulting content set. If you don't pass the flag you should get a record per each element.

**Reply by paprophet** [reply]
thank you!
how to  the unit change?
for example a property ( property name is 'Total Thickness' ) of the element  value is '24' and  'unit' is m ,and displayvalue is '2400 mm'
so what method in imodeljs can change this (24 m to 2400 mm) ?

**Reply by grigasp** [reply]
Properties in ECSchemas define their kind of quantity, which in turn defines it's persistence and supported presentation units. From the information you gave it looks like 'm' is the persistence unit and presentation unit is 'mm'. 

Currently we don't provide a way to choose between different presentation units of the same unit system when requesting content - the first unit of requested unit system is used. This means you would have to modify ECSchema to change presentation unit for the property from 'mm' to 'm' or something else in metric system. We do, however, allow choosing between different unit systems (imperial, metric, etc.), so if the property supports both imperial and metric, you can switch between them by passing appropriate value for the `unitSystem` attribute for content request props.

@ColinKerr, got anything more to add?
